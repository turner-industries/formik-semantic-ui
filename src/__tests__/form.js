import React from 'react';
import {render, cleanup, fireEvent, wait} from 'react-testing-library';

import {findAndClick, setFieldValue} from '../test-utils';
import {Button, Form} from '../index';
import FormikInput from '../Input';

describe('Form', () => {
  afterEach(() => {
    cleanup();
  });

  it('child as a function', async () => {
    const handleSubmit = jest.fn();

    const {container, getByText} = render(
      <Form
        initialValues={{name: ''}}
        onSubmit={handleSubmit}
        data-testid="test-id"
      >
        {({isSubmitting}) => (
          <React.Fragment>
            <Button.Submit disabled={isSubmitting}>Submit</Button.Submit>
          </React.Fragment>
        )}
      </Form>
    );

    const submitButton = getByText('Submit');

    await fireEvent.click(submitButton);
    await fireEvent.click(submitButton);

    await wait(() => {
      expect(handleSubmit).toHaveBeenCalled();
    });

    expect(handleSubmit.mock.calls.length).toEqual(1);
    expect(container).toMatchSnapshot();
  });

  it('render props', async () => {
    const {container, getByText} = render(
      <Form
        initialValues={{name: ''}}
        onSubmit={() => {}}
        render={({isSubmitting}) => (
          <React.Fragment>
            <Button.Submit disabled={isSubmitting}>Submit</Button.Submit>
          </React.Fragment>
        )}
      />
    );

    const submitButton = getByText('Submit');
    await fireEvent.click(submitButton);
    expect(container).toMatchSnapshot();
  });

  it('ignore loading', async () => {
    const {container, getByText} = render(
      <Form
        initialValues={{name: ''}}
        ignoreLoading
        onSubmit={() => {}}
        render={({isSubmitting}) => (
          <React.Fragment>
            <Button.Submit disabled={isSubmitting}>Submit</Button.Submit>
          </React.Fragment>
        )}
      />
    );

    const submitButton = getByText('Submit');
    await fireEvent.click(submitButton);
    expect(container).toMatchSnapshot();
  });

  it('server validation', async () => {
    let isTouched;
    const {getByText, getByLabelText} = render(
      <Form
        initialValues={{name: ''}}
        onSubmit={() => {}}
        serverValidation
        render={({isSubmitting, touched}) => {
          isTouched = touched;
          return (
            <React.Fragment>
              <FormikInput name="name" label="Name" />
              <Button.Submit disabled={isSubmitting}>Submit</Button.Submit>
            </React.Fragment>
          );
        }}
      />
    );

    // make field dirty
    setFieldValue(getByLabelText(/name/i), 'test');
    expect(isTouched).toEqual({name: true});

    await findAndClick(() => getByText('Submit'));

    // field should reset after submit
    await wait(() => {
      expect(isTouched).toEqual({name: false});
    });
  });

  it('errors in complex types', async () => {
    const {getByText, container} = render(
      <Form
        initialValues={{values: [{ name: 'test' }]}}
        validate={() => ({ values: [{ name: 'Error' }]})}
        onSubmit={() => {}}
        render={() => {
          return (
            <React.Fragment>
              <FormikInput name="values.0.name" label="Name" />
              <Button.Submit>Submit</Button.Submit>
            </React.Fragment>
          );
        }}
      />
    );

    await findAndClick(() => getByText('Submit'));

    expect(container).toMatchSnapshot();
  });

  describe('Schema', () => {
    it('Input', () => {
      const schema = {
        name: {
          type: 'text',
          label: 'Name',
        },
      };

      const {container} = render(
        <Form initialValues={{name: ''}} onSubmit={() => {}} schema={schema}>
          <Button.Submit>Submit</Button.Submit>
        </Form>
      );

      expect(container).toMatchSnapshot();
    });

    it('Input with value', () => {
      const schema = {
        name: {
          type: 'text',
          label: 'Name',
          value: 'Justin',
        },
      };

      const {container} = render(
        <Form initialValues={{name: ''}} onSubmit={() => {}} schema={schema}>
          <Button.Submit>Submit</Button.Submit>
        </Form>
      );

      expect(container).toMatchSnapshot();
    });

    it('Input: inputProps & fieldProps', () => {
      const schema = {
        ssn: {
          type: 'password',
          label: 'SSN',
          inputProps: {
            placeholder: 'SSN',
          },
          fieldProps: {
            width: '8',
          },
        },
      };

      const {container} = render(
        <Form initialValues={{name: ''}} onSubmit={() => {}} schema={schema}>
          <Button.Submit>Submit</Button.Submit>
        </Form>
      );

      expect(container).toMatchSnapshot();
    });

    it('TextArea', () => {
      const schema = {
        notes: {
          type: 'textarea',
          label: 'Notes',
        },
      };

      const {container} = render(
        <Form initialValues={{name: ''}} onSubmit={() => {}} schema={schema}>
          <Button.Submit>Submit</Button.Submit>
        </Form>
      );

      expect(container).toMatchSnapshot();
    });

    it('Checkbox', () => {
      const schema = {
        agree: {
          type: 'checkbox',
          label: 'I Agree',
        },
      };

      const {container} = render(
        <Form initialValues={{name: ''}} onSubmit={() => {}} schema={schema}>
          <Button.Submit>Submit</Button.Submit>
        </Form>
      );

      expect(container).toMatchSnapshot();
    });

    it('Dropdown', () => {
      const schema = {
        name: {
          type: 'dropdown',
          label: 'Name',
          options: [
            {text: 'Justin', value: 'justin'},
            {text: 'Not Justin', value: 'not-justin'},
          ],
        },
      };

      const {container} = render(
        <Form initialValues={{name: ''}} onSubmit={() => {}} schema={schema}>
          <Button.Submit>Submit</Button.Submit>
        </Form>
      );

      expect(container).toMatchSnapshot();
    });

    it('Dropdown: allow additions', async  () => {
      const schema = {
        name: {
          type: 'dropdown',
          label: 'Name',
          options: [
            {text: 'Justin', value: 'justin', key: 'justin'},
            {text: 'Not Justin', value: 'not-justin', key: 'not-justin'},
          ],
          inputProps: {
            multiple: true,
            selection: true,
            search: true,
            allowAdditions: true
          }
        },
      };

      const {container, getByText, getByLabelText} = render(
        <Form initialValues={{name: ''}} onSubmit={() => {}} schema={schema}>
          <Button.Submit>Submit</Button.Submit>
        </Form>
      );

      const input = getByLabelText(/Name/i).querySelector('input')

      setFieldValue(input, ['Paula']);
      fireEvent.keyDown(input, { key: 'Enter', code: 13 })

      expect(container).toMatchSnapshot();
    });
  });
});
