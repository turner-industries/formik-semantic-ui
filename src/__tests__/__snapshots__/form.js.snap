// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Schema Checkbox 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="field"
    >
      <div
        class="ui checkbox"
      >
        <input
          class="hidden"
          id="field_checkbox_agree"
          name="agree"
          readonly=""
          tabindex="0"
          type="checkbox"
          value=""
        />
        <label
          for="field_checkbox_agree"
        >
          I Agree
        </label>
      </div>
    </div>
    <button
      type="submit"
      class="ui primary button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form Schema Dropdown 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="field"
    >
      <label
        for="field_dropdown_name"
      >
        Name
      </label>
      <div
        id="field_dropdown_name"
        name="name"
        role="listbox"
        aria-expanded="false"
        class="ui selection dropdown"
        tabindex="0"
      >
        <div
          class="text"
          role="alert"
          aria-live="polite"
        />
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
        >
          <div
            style="pointer-events: all;"
            role="option"
            aria-checked="false"
            aria-selected="true"
            class="selected item"
          >
            <span
              class="text"
            >
              Justin
            </span>
          </div>
          <div
            style="pointer-events: all;"
            role="option"
            aria-checked="false"
            aria-selected="false"
            class="item"
          >
            <span
              class="text"
            >
              Not Justin
            </span>
          </div>
        </div>
      </div>
    </div>
    <button
      type="submit"
      class="ui primary button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form Schema Dropdown: allow additions 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="field"
    >
      <label
        for="field_dropdown_name"
      >
        Name
      </label>
      <div
        id="field_dropdown_name"
        name="name"
        role="combobox"
        aria-expanded="true"
        class="ui active visible multiple search selection dropdown"
      >
        <a
          class="ui label"
          value="Paula"
        >
          Paula
          <i
            aria-hidden="true"
            class="delete icon"
          />
        </a>
        <input
          aria-autocomplete="list"
          autocomplete="off"
          class="search"
          tabindex="0"
          type="text"
          value=""
          style=""
        />
        <span
          class="sizer"
          style=""
        >
          Paula
        </span>
        <div
          class="text"
          role="alert"
          aria-live="polite"
        />
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          aria-multiselectable="true"
          role="listbox"
          class="visible menu transition"
        >
          <div
            style="pointer-events: all;"
            role="option"
            aria-checked="false"
            aria-selected="true"
            class="selected item"
          >
            <span
              class="text"
            >
              Justin
            </span>
          </div>
          <div
            style="pointer-events: all;"
            role="option"
            aria-checked="false"
            aria-selected="false"
            class="item"
          >
            <span
              class="text"
            >
              Not Justin
            </span>
          </div>
        </div>
      </div>
    </div>
    <button
      type="submit"
      class="ui primary button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form Schema Input 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="field"
    >
      <label
        for="field_input_name"
      >
        Name
      </label>
      <div
        class="ui input"
      >
        <input
          id="field_input_name"
          name="name"
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      type="submit"
      class="ui primary button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form Schema Input with value 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="field"
    >
      <label
        for="field_input_name"
      >
        Name
      </label>
      <div
        class="ui input"
      >
        <input
          id="field_input_name"
          name="name"
          type="text"
          value="Justin"
        />
      </div>
    </div>
    <button
      type="submit"
      class="ui primary button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form Schema Input: inputProps & fieldProps 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="eight wide field"
    >
      <label
        for="field_input_ssn"
      >
        SSN
      </label>
      <div
        class="ui input"
      >
        <input
          id="field_input_ssn"
          name="ssn"
          placeholder="SSN"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      type="submit"
      class="ui primary button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form Schema TextArea 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="field"
    >
      <label
        for="field_textarea_notes"
      >
        Notes
      </label>
      <textarea
        id="field_textarea_notes"
        name="notes"
        rows="4"
      />
    </div>
    <button
      type="submit"
      class="ui primary button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form child as a function 1`] = `
<div>
  <form
    data-testid="test-id"
    class="ui loading form"
  >
    <button
      type="submit"
      class="ui primary disabled button"
      disabled=""
      tabindex="-1"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form errors in complex types 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="error field"
    >
      <label
        for="field_input_values.0.name"
      >
        Name
      </label>
      <div
        class="ui input"
      >
        <input
          id="field_input_values.0.name"
          name="values.0.name"
          type="text"
          value="test"
        />
      </div>
      <span
        class="sui-error-message"
      >
        Error
      </span>
    </div>
    <button
      type="submit"
      class="ui primary button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form ignore loading 1`] = `
<div>
  <form
    class="ui form"
  >
    <button
      type="submit"
      class="ui primary disabled button"
      disabled=""
      tabindex="-1"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form render props 1`] = `
<div>
  <form
    class="ui loading form"
  >
    <button
      type="submit"
      class="ui primary disabled button"
      disabled=""
      tabindex="-1"
    >
      Submit
    </button>
  </form>
</div>
`;
